# Mount dataset resources

In [1]:
import re
import pandas as pd
import numpy as np
import warnings

import datetime
from datetime import datetime, date
import time

# Visualisation
from matplotlib import pyplot as plt

# Suppress warnings
warnings.filterwarnings("ignore")

# чтениие/запись файлов в feather-формате
import feather

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [3]:
%cd '/content/drive/MyDrive/github/eda_examples/predictive-maintenance/'
%pwd

/content/drive/MyDrive/github/eda_examples/predictive-maintenance


'/content/drive/MyDrive/github/eda_examples/predictive-maintenance'

# Data Wrangling

In [4]:
import pandas as pd
import numpy as np


### Data Source ###

___Training Data:___  The aircraft engine run-to-failure data.
[download trianing data](http://azuremlsamples.azureml.net/templatedata/PM_train.txt)  
___Test Data:___ The aircraft engine operating data without failure events recorded.
[download test data](http://azuremlsamples.azureml.net/templatedata/PM_test.txt)  
___Ground Truth Data:___ The true remaining cycles for each engine in the testing data.
[download truth data](http://azuremlsamples.azureml.net/templatedata/PM_truth.txt)  

For simplicity, data files have been downloaded to local Data folder.

### Data Columns

•	__id__: is the engine ID, ranging from 1 to 100  
•	__cycle__: per engine sequence, starts from 1 to the cycle number where failure had happened (trining data only) 
•	__setting1__ to __setting3__: engine operational settings  
•	__s1__ to __s21__: sensors measurements  


In [5]:
#dataset column names:

col_names = ['id','cycle','setting1','setting2','setting3','s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']


### Load Training Data:

In [6]:
#load training data

df_train_raw = pd.read_csv('./data/PM_train.txt', sep = ' ', header=None)
df_train_raw.head()

0   1       2       3      4       5   ...    22     23     24       25  26  27
0   1   1 -0.0007 -0.0004  100.0  518.67  ...  2388  100.0  39.06  23.4190 NaN NaN
1   1   2  0.0019 -0.0003  100.0  518.67  ...  2388  100.0  39.00  23.4236 NaN NaN
2   1   3 -0.0043  0.0003  100.0  518.67  ...  2388  100.0  38.95  23.3442 NaN NaN
3   1   4  0.0007  0.0000  100.0  518.67  ...  2388  100.0  38.88  23.3739 NaN NaN
4   1   5 -0.0019 -0.0002  100.0  518.67  ...  2388  100.0  38.90  23.4044 NaN NaN

[5 rows x 28 columns]

In [8]:
#drop extra space columnn

df_train_raw.drop([26,27], axis=1, inplace=True)

In [9]:
#assign column names

df_train_raw.columns = col_names
df_train_raw.head()

id  cycle  setting1  setting2  setting3  ...  s17   s18    s19    s20      s21
0   1      1   -0.0007   -0.0004     100.0  ...  392  2388  100.0  39.06  23.4190
1   1      2    0.0019   -0.0003     100.0  ...  392  2388  100.0  39.00  23.4236
2   1      3   -0.0043    0.0003     100.0  ...  390  2388  100.0  38.95  23.3442
3   1      4    0.0007    0.0000     100.0  ...  392  2388  100.0  38.88  23.3739
4   1      5   -0.0019   -0.0002     100.0  ...  393  2388  100.0  38.90  23.4044

[5 rows x 26 columns]

In [10]:
# get some stat

df_train_raw.describe()

id         cycle  ...           s20           s21
count  20631.000000  20631.000000  ...  20631.000000  20631.000000
mean      51.506568    108.807862  ...     38.816271     23.289705
std       29.227633     68.880990  ...      0.180746      0.108251
min        1.000000      1.000000  ...     38.140000     22.894200
25%       26.000000     52.000000  ...     38.700000     23.221800
50%       52.000000    104.000000  ...     38.830000     23.297900
75%       77.000000    156.000000  ...     38.950000     23.366800
max      100.000000    362.000000  ...     39.430000     23.618400

[8 rows x 26 columns]

There are 100 engines. each engine has between 1 to 362 cycles (average of 108 cycles per engine). The last cycle for each engine represents the cycle when failure had happened.

In [11]:
# check the data types

df_train_raw.dtypes

id            int64
cycle         int64
setting1    float64
setting2    float64
setting3    float64
s1          float64
s2          float64
s3          float64
s4          float64
s5          float64
s6          float64
s7          float64
s8          float64
s9          float64
s10         float64
s11         float64
s12         float64
s13         float64
s14         float64
s15         float64
s16         float64
s17           int64
s18           int64
s19         float64
s20         float64
s21         float64
dtype: object

All data columns are numeric.

In [12]:
# check for NaN values

df_train_raw.isnull().sum()

id          0
cycle       0
setting1    0
setting2    0
setting3    0
s1          0
s2          0
s3          0
s4          0
s5          0
s6          0
s7          0
s8          0
s9          0
s10         0
s11         0
s12         0
s13         0
s14         0
s15         0
s16         0
s17         0
s18         0
s19         0
s20         0
s21         0
dtype: int64

No missing values. This is a clean dataset!

### Load Test Data:

In [13]:
#load test data

df_test_raw = pd.read_csv('./data/PM_test.txt', sep = ' ', header=None)
df_test_raw.head()


0   1       2       3      4       5   ...    22     23     24       25  26  27
0   1   1  0.0023  0.0003  100.0  518.67  ...  2388  100.0  38.86  23.3735 NaN NaN
1   1   2 -0.0027 -0.0003  100.0  518.67  ...  2388  100.0  39.02  23.3916 NaN NaN
2   1   3  0.0003  0.0001  100.0  518.67  ...  2388  100.0  39.08  23.4166 NaN NaN
3   1   4  0.0042  0.0000  100.0  518.67  ...  2388  100.0  39.00  23.3737 NaN NaN
4   1   5  0.0014  0.0000  100.0  518.67  ...  2388  100.0  38.99  23.4130 NaN NaN

[5 rows x 28 columns]

In [14]:
#drop extra space columnn
df_test_raw.drop([26,27], axis=1, inplace=True)

#assign column names
df_test_raw.columns = col_names
df_test_raw.head()


id  cycle  setting1  setting2  setting3  ...  s17   s18    s19    s20      s21
0   1      1    0.0023    0.0003     100.0  ...  392  2388  100.0  38.86  23.3735
1   1      2   -0.0027   -0.0003     100.0  ...  393  2388  100.0  39.02  23.3916
2   1      3    0.0003    0.0001     100.0  ...  393  2388  100.0  39.08  23.4166
3   1      4    0.0042    0.0000     100.0  ...  391  2388  100.0  39.00  23.3737
4   1      5    0.0014    0.0000     100.0  ...  390  2388  100.0  38.99  23.4130

[5 rows x 26 columns]

In [15]:
# get some stat on test data

df_test_raw.describe()

id         cycle  ...           s20           s21
count  13096.000000  13096.000000  ...  13096.000000  13096.000000
mean      51.543907     76.836515  ...     38.892502     23.335743
std       28.289423     53.057749  ...      0.141681      0.084120
min        1.000000      1.000000  ...     38.310000     22.935400
25%       28.000000     33.000000  ...     38.800000     23.281600
50%       52.000000     69.000000  ...     38.900000     23.339200
75%       76.000000    113.000000  ...     38.990000     23.393600
max      100.000000    303.000000  ...     39.410000     23.641900

[8 rows x 26 columns]

Same as training data, there are 100 engines, each engine has between 1 to 303 cycles (average of 76 cycles per engine). But this time, failure cycle was not provided.  

Failure events for test data - remaining cycles before failure (TTF) - were provided in a separate truth file.

### Load Truth Data:

In [16]:
# Load the truth data - actual 'ttf' for test data

df_truth = pd.read_csv('./data/PM_truth.txt', sep = ' ', header=None)
df_truth.head()

0   1
0  112 NaN
1   98 NaN
2   69 NaN
3   82 NaN
4   91 NaN

In [17]:
#drop extra empty column in the truth data and rename remaining 'ttf'

df_truth.drop([1], axis=1, inplace=True)
df_truth.columns = ['ttf']
df_truth.head()

ttf
0  112
1   98
2   69
3   82
4   91

In [18]:
#get some stat on truth data

df_truth.describe()

ttf
count  100.00000
mean    75.52000
std     41.76497
min      7.00000
25%     32.75000
50%     86.00000
75%    112.25000
max    145.00000

To get meaningful test data, we need to merge the truth data (TTF) with last cycle for each engine in the test data. This will give us a test set of 100 engines with their TTF data. Will do that later when we create regression and classification labels for both training and test data. 

But now let us add some features to smooth the sensors reading: rolling average and rolling standard deviation.


### Feature Extraction:

Create helper function to create features based on smoothing the time series for sensors by adding rolling mean and rolling standard deviation

In [27]:

def add_features(df_in, rolling_win_size):
    
    """Add rolling average and rolling standard deviation for sensors readings using fixed rolling window size.
    
    Args:
            df_in (dataframe)     : The input dataframe to be proccessed (training or test) 
            rolling_win_size (int): The window size, number of cycles for applying the rolling function
        
    Reurns:
            dataframe: contains the input dataframe with additional rolling mean and std for each sensor
    
    """
    
    sensor_cols = ['s1','s2','s3','s4','s5','s6','s7','s8','s9','s10','s11','s12','s13','s14','s15','s16','s17','s18','s19','s20','s21']
    
    sensor_av_cols = [nm.replace('s', 'av') for nm in sensor_cols]
    sensor_sd_cols = [nm.replace('s', 'sd') for nm in sensor_cols]
    
    df_out = pd.DataFrame()
    
    ws = rolling_win_size
    
    #calculate rolling stats for each engine id
    
    for m_id in pd.unique(df_in.id):
    
        # get a subset for each engine sensors
        df_engine = df_in[df_in['id'] == m_id]
        df_sub = df_engine[sensor_cols]

    
        # get rolling mean for the subset
        av = df_sub.rolling(ws, min_periods=1).mean()
        av.columns = sensor_av_cols
    
        # get the rolling standard deviation for the subset
        sd = df_sub.rolling(ws, min_periods=1).std().fillna(0)
        sd.columns = sensor_sd_cols
    
        # combine the two new subset dataframes columns to the engine subset
        new_ftrs = pd.concat([df_engine,av,sd], axis=1)
    
        # add the new features rows to the output dataframe
        df_out = pd.concat([df_out,new_ftrs])
        
    return df_out

create helper function to add the regression and classification labels to the training data

In [28]:

def prepare_train_data (df_in, period):
    
    """Add regression and classification labels to the training data.

        Regression label: ttf (time-to-failure) = each cycle# for an engine subtracted from the last cycle# of the same engine
        Binary classification label: label_bnc = if ttf is <= parameter period then 1 else 0 (values = 0,1)
        Multi-class classification label: label_mcc = 2 if ttf <= 0.5* parameter period , 1 if ttf<= parameter period, else 2
        
      Args:
          df_in (dataframe): The input training data
          period (int)     : The number of cycles for TTF segmentation. Used to derive classification labels
          
      Returns:
          dataframe: The input dataframe with regression and classification labels added
          
    """
    
    #create regression label
    
    #make a dataframe to hold the last cycle for each enginge in the dataset
    df_max_cycle = pd.DataFrame(df_in.groupby('id')['cycle'].max())
    df_max_cycle.reset_index(level=0, inplace=True)
    df_max_cycle.columns = ['id', 'last_cycle']
    
    #add time-to-failure ttf as a new column - regression label
    df_in = pd.merge(df_in, df_max_cycle, on='id')
    df_in['ttf'] = df_in['last_cycle'] - df_in['cycle']
    df_in.drop(['last_cycle'], axis=1, inplace=True)
    
    #create binary classification label
    df_in['label_bnc'] = df_in['ttf'].apply(lambda x: 1 if x <= period else 0)
    
    #create multi-class classification label
    df_in['label_mcc'] = df_in['ttf'].apply(lambda x: 2 if x <= period/2 else 1 if x <= period else 0)
    
    return df_in
    

create helper function to add the regression and classification labels to the training data

In [29]:

def prepare_test_data(df_test_in, df_truth_in, period):
    
    """Add regression and classification labels to the test data.

        Regression label: ttf (time-to-failure) = extract the last cycle for each enginge and then merge the record with the truth data
        Binary classification label: label_bnc = if ttf is <= parameter period then 1 else 0 (values = 0,1)
        Multi-class classification label: label_mcc = 2 if ttf <= 0.5* parameter period , 1 if ttf<= parameter period, else 2
        
      Args:
          df_in (dataframe): The input training data
          period (int)     : The number of cycles for TTF segmentation. Used to derive classification labels
          
      Returns:
          dataframe: The input dataframe with regression and classification labels added
    

    
    """
    
    df_tst_last_cycle = pd.DataFrame(df_test_in.groupby('id')['cycle'].max())
    
    df_tst_last_cycle.reset_index(level=0, inplace=True)
    df_tst_last_cycle.columns = ['id', 'last_cycle']
    
    df_test_in = pd.merge(df_test_in, df_tst_last_cycle, on='id')


    df_test_in = df_test_in[df_test_in['cycle'] == df_test_in['last_cycle']]

    df_test_in.drop(['last_cycle'], axis=1, inplace=True)
    
    df_test_in.reset_index(drop=True, inplace=True)
    
    df_test_in = pd.concat([df_test_in, df_truth], axis=1)
    
    #create binary classification label
    df_test_in['label_bnc'] = df_test_in['ttf'].apply(lambda x: 1 if x <= period else 0)
    
    #create multi-class classification label
    df_test_in['label_mcc'] = df_test_in['ttf'].apply(lambda x: 2 if x <= period/2 else 1 if x <= period else 0)

    return df_test_in

With the help of these functions, let us prepare training and test data by adding features and labels

### Prepare the Training Data:

In [30]:
# add extracted features to training data

df_train_fx = add_features(df_train_raw, 5)
df_train_fx.head()

id  cycle  setting1  setting2  ...  sd18  sd19      sd20      sd21
0   1      1   -0.0007   -0.0004  ...   0.0   0.0  0.000000  0.000000
1   1      2    0.0019   -0.0003  ...   0.0   0.0  0.042426  0.003253
2   1      3   -0.0043    0.0003  ...   0.0   0.0  0.055076  0.044573
3   1      4    0.0007    0.0000  ...   0.0   0.0  0.076322  0.037977
4   1      5   -0.0019   -0.0002  ...   0.0   0.0  0.073621  0.033498

[5 rows x 68 columns]

In [31]:
#add labels to training data using period of 30 cycles for classification

df_train = prepare_train_data (df_train_fx, 30)
df_train.head()

id  cycle  setting1  setting2  ...      sd21  ttf  label_bnc  label_mcc
0   1      1   -0.0007   -0.0004  ...  0.000000  191          0          0
1   1      2    0.0019   -0.0003  ...  0.003253  190          0          0
2   1      3   -0.0043    0.0003  ...  0.044573  189          0          0
3   1      4    0.0007    0.0000  ...  0.037977  188          0          0
4   1      5   -0.0019   -0.0002  ...  0.033498  187          0          0

[5 rows x 71 columns]

In [32]:
df_train.dtypes

id             int64
cycle          int64
setting1     float64
setting2     float64
setting3     float64
              ...   
sd20         float64
sd21         float64
ttf            int64
label_bnc      int64
label_mcc      int64
Length: 71, dtype: object

Rolling average, rolling standard deviation, regression labels, and classification labels have been added to the training data.  

Let us save the dataframe for later use in data exploration and modeling phases.

In [33]:
# save the training data to csv file for later use

df_train.to_csv('./data/train.csv', index=False)

### Prepare the Test Data:

In [34]:
# add extracted features to test data

df_test_fx = add_features(df_test_raw, 5)
df_test_fx.head()

id  cycle  setting1  setting2  ...  sd18  sd19      sd20      sd21
0   1      1    0.0023    0.0003  ...   0.0   0.0  0.000000  0.000000
1   1      2   -0.0027   -0.0003  ...   0.0   0.0  0.113137  0.012799
2   1      3    0.0003    0.0001  ...   0.0   0.0  0.113725  0.021642
3   1      4    0.0042    0.0000  ...   0.0   0.0  0.093095  0.020353
4   1      5    0.0014    0.0000  ...   0.0   0.0  0.080623  0.020672

[5 rows x 68 columns]

In [35]:
#add labels to test data using period of 30 cycles for classification

df_test = prepare_test_data(df_test_fx, df_truth, 30)
df_test.head()

id  cycle  setting1  setting2  ...      sd21  ttf  label_bnc  label_mcc
0   1     31   -0.0006    0.0004  ...  0.025037  112          0          0
1   2     49    0.0018   -0.0001  ...  0.027820   98          0          0
2   3    126   -0.0016    0.0004  ...  0.018966   69          0          0
3   4    106    0.0012    0.0004  ...  0.021106   82          0          0
4   5     98   -0.0013   -0.0004  ...  0.101100   91          0          0

[5 rows x 71 columns]

In [36]:
df_test.dtypes

id             int64
cycle          int64
setting1     float64
setting2     float64
setting3     float64
              ...   
sd20         float64
sd21         float64
ttf            int64
label_bnc      int64
label_mcc      int64
Length: 71, dtype: object

Rolling average, rolling standard deviation, regression labels, and classification labels have been added to the test data.  

Let us save the dataframe for later use in data exploration and modeling phases

In [37]:
# save the test data to csv file for later use

df_test.to_csv('./data/test.csv', index=False)